File: 36 Valid Sudoku/Solution.java
import java.util.HashSet;

class Solution36 {
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character>[] rowSet = new HashSet[9];
        HashSet<Character>[] colSet = new HashSet[9];
        HashSet<Character>[] squareSet = new HashSet[9];

        // Initialize the sets for rows, columns, and squares
        for (int i = 0; i < 9; i++) {
            rowSet[i] = new HashSet<>();
            colSet[i] = new HashSet<>();
            squareSet[i] = new HashSet<>();
        }

        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                char value = board[row][col];

                if (value == '.') {
                    continue; // Skip empty cells
                }

                // Check row, column, and 3x3 square for duplicates
                if (rowSet[row].contains(value) || colSet[col].contains(value) || squareSet[(row / 3) * 3 + (col / 3)].contains(value)) {
                    return false;
                }

                // Add the value to the corresponding sets
                rowSet[row].add(value);
                colSet[col].add(value);
                squareSet[(row / 3) * 3 + (col / 3)].add(value);
            }
        }
        return true; // The board is valid
    }

    public static void main(String[] args) {
        Solution36 solution = new Solution36();

        char[][] board = {
                {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                {'.', '.', '.', '.', '8', '.', '.', '7', '9'}
        };

        boolean result = solution.isValidSudoku(board);
        System.out.println(result);  // Should print true or false
    }
}

File: 36 Valid Sudoku/Solution.py
'''
Time Complexity: O(1) or O(n) where n is row*column which is indeed 81
Space Complexity: O(3n) Every element is part of row/col/sqw

if not using defaultdict(set) can use below
rows, columns, square = {}, {}, {}
if r not in rows:
    rows[r] = set()
if c not in columns:
    columns[c] = set()
if (r//3, c//3) not in square:
    square[(r//3, c//3)] = set()
'''
from collections import defaultdict

class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        row_dict = defaultdict(set)
        col_dict = defaultdict(set)
        sqr_dict = defaultdict(set)

        for row in range(9):
            for col in range(9):
                if board[row][col] == ".":
                    continue
                if board[row][col] in row_dict[row] or board[row][col] in col_dict[col] or board[row][col] in sqr_dict[(row//3,col//3)]:
                    return False
                row_dict[row].add(board[row][col])
                col_dict[col].add(board[row][col])
                sqr_dict[(row//3,col//3)].add(board[row][col])
        return True

if __name__ == "__main__":
    solution = Solution()
    board = [["5", "3", ".", ".", "7", ".", ".", ".", "."]
           , ["6", ".", ".", "1", "9", "5", ".", ".", "."]
           , [".", "9", "8", ".", ".", ".", ".", "6", "."]
           , ["8", ".", ".", ".", "6", ".", ".", ".", "3"]
           , ["4", ".", ".", "8", ".", "3", ".", ".", "1"]
           , ["7", ".", ".", ".", "2", ".", ".", ".", "6"]
           , [".", "6", ".", ".", ".", ".", "2", "8", "."]
           , [".", ".", ".", "4", "1", "9", ".", ".", "5"]
           , [".", ".", ".", ".", "8", ".", ".", "7", "9"]]
    board1 =[["8", "3", ".", ".", "7", ".", ".", ".", "."]
           , ["6", ".", ".", "1", "9", "5", ".", ".", "."]
           , [".", "9", "8", ".", ".", ".", ".", "6", "."]
           , ["8", ".", ".", ".", "6", ".", ".", ".", "3"]
           , ["4", ".", ".", "8", ".", "3", ".", ".", "1"]
           , ["7", ".", ".", ".", "2", ".", ".", ".", "6"]
           , [".", "6", ".", ".", ".", ".", "2", "8", "."]
           , [".", ".", ".", "4", "1", "9", ".", ".", "5"]
           , [".", ".", ".", ".", "8", ".", ".", "7", "9"]]
    result = solution.isValidSudoku(board)
    print(result)